### PR TITLE
Packing<tuple> specialization

### PR DESCRIPTION
**Pull request by roystgnr**
This gives us a Packing specialization for any tuple that's not completely composed of fixed types, and removes those cases from the previous StandardType specialization, so variable-size data structures with tuples in them can now participate in automatic packing dispatch.

Hoping @lindsayad can look this over too; he did much of the hard work this is based on.

And, if it's not too late, @loganharbour might want to try this out, since he was the first one to try to communicate a vector of tuples with strings in them under the naive theory that just because we supported vectors and tuples and strings that would mean we must support vectors of tuples of strings too.

### CODE REVIEW

**Comment by lindsayad** [comment]
I'll look at this after department meeting!

File: src/parallel/include/timpi/packing.h [hunk 1]
```diff
@@ -284,6 +285,263 @@ Packing<std::pair<T1, T2>,
...
+            typename Context,
+            typename std::enable_if<!IsFixed<T3>::value, int>::type = 0>
+  static unsigned int packable_size_comp(const T3 & comp, const Context * ctx)
+  {
+    return Packing<T3>::packable_size(comp, ctx);
+  }
+
+  template <typename Context,
+            std::size_t I>
+  static typename std::enable_if<I == sizeof...(Types), unsigned int>::type
```
Comment (lindsayad) [inline src/parallel/include/timpi/packing.h:384]: Nice

**Review by lindsayad — APPROVED** [review]
This is awesome

File: src/parallel/include/timpi/packing.h (hunk 1 above)
Comment (roystgnr) [reply src/parallel/include/timpi/packing.h:384]: Wish I could take credit for it.  I did a search for something like "how to iterate over a tuple", and felt like an idiot once I saw the phrase "tail recursion".  I never took a real CS class, but I have learned the basics, and I have used them before (for the heterogeneous structures in MetaPhysicL), and I shouldn't have needed reminding about how.

**Pull request merged**